python/pfs/pipe2d/generateCommands.py

```diff
@@ -0,0 +1,1493 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+"""Components to implement a program ``generateCommands.py``.
+"""
+
+__all__ = []  # modified by @export
+
+import lsst.log
+
+import argparse
+import contextlib
+import dataclasses
+import enum
+import os
+import re
+import shlex
+import textwrap
+
+import yaml
+
+from typing import Any, Dict, Iterable, List, Mapping, Optional, TextIO, Tuple, Type, Union
+
+
+def export(obj):
+    """Decorator to add obj to __all__.
+    """
+    __all__.append(obj.__name__)
+    return obj
+
+
+DEFAULT_CALIB_VALIDITY = 1800
+"""Calibs' valid days, used if 'validity' field is not present in YAML files.
+"""
+
+
+@export
+def main():
+    """Parse ``sys.argv`` and call ``generateCommands()``.
+    """
+    parser = argparse.ArgumentParser(description="""
+    Process a yaml file and generate shell commands.
+    """)
+
+    parser.add_argument("dataDir", type=str, help="Root of data repository")
+    parser.add_argument("specFile", type=str, help="Name of file specifying the work")
+    parser.add_argument("--init", action="store_true", help="Install initial calibs")
+    parser.add_argument("--blocks", type=str, nargs="+", help="Blocks to execute")
+    parser.add_argument("--calib", type=str, default="CALIB", help="Name of output calibration directory")
+    parser.add_argument("--calibTypes", type=str, nargs="+", default=[], choices=CalibBlock.calibTypes,
+                        help="Types of calibs to process")
+    parser.add_argument("--clean", action="store_true",
+                        help="Clean up byproducts after ingesting calibs")
+    parser.add_argument("--copyMode", choices=["move", "copy", "link", "skip"], default="copy",
+                        help="How to move files into calibration directory")
+    parser.add_argument("--devel", action="store_true",
+                        help="Run commands in the development mode (no versioning)")
+    parser.add_argument("--force", action="store_true",
+                        help="Continue in the face of problems")
+    parser.add_argument("-j", "--processes", type=int, default=1, help="Number of processes to use")
+    parser.add_argument("-L", "--loglevel", type=str, choices=list(LogLevel.__members__), default="INFO",
+                        help="How chatty should I be?")
+    parser.add_argument("-o", "--output", type=str, default="a.sh", help="Path to output file")
```

Comment:
Can we just make this a required argument, instead of an obscure (but not without precedent, I know) default?

Reply:
I have promoted `--output` to the third positional parameter. I could keep it an option parameter, just making it required, but the document of `argparse` says it is bad mannered.